cassiopeia/solver/UPGMASolver.py

```diff
@@ -0,0 +1,213 @@
+    ):
+
+        super().__init__(
+            dissimilarity_function=dissimilarity_function,
+            add_root=True,
+            prior_transformation=prior_transformation,
+        )
+
+
+    def root_tree(self, tree: nx.Graph, root_sample: str, remaining_samples: List[str]):
+        """Roots the tree.
+
+        Adds the root at the top of the UPGMA reconstructed tree. By the 
+        ultrametric assumption, the root is placed as the parent to the last 
+        two unjoined nodes.
+
+        Args:
+            tree: Networkx object representing the tree topology
+            root_sample: Ignored in this case, the root is known in this case
+            remaining_samples: The last two unjoined nodes in the tree
+
+        Returns:
+            A rooted tree.
+        """
+
+        tree.add_node("root")
+        tree.add_edges_from([("root", remaining_samples[0]), ("root", remaining_samples[1])])
+        
+        rooted_tree = nx.DiGraph()
+        for e in nx.dfs_edges(tree, source="root"):
+            rooted_tree.add_edge(e[0], e[1])
+
+        return rooted_tree
+
+
+    def find_cherry(self, dissimilarity_matrix: np.array) -> Tuple[int, int]:
+        """Finds a pair of samples to join into a cherry.
+
+        Finds the pair of samples with the minimum dissimilarity by finding the
+        minimum value in the provided dissimilarity matrix
+
+        Args:
+            dissimilarity_matrix: A sample x sample dissimilarity matrix
+
+        Returns:
+            A tuple of integers representing rows in the dissimilarity matrix
+                to join.
+        """
+
+        dissimilarity_matrix = dissimilarity_matrix.astype(float)
+        np.fill_diagonal(dissimilarity_matrix, np.inf)
+
+        return np.unravel_index(np.argmin(dissimilarity_matrix, axis=None), dissimilarity_matrix.shape)
+
+
+    def update_dissimilarity_map(
+        self,
+        dissimilarity_map: pd.DataFrame,
+        cherry: Tuple[str, str],
+        new_node: str,
+        cluster_to_cluster_size: Dict[str, int]
+    ) -> pd.DataFrame:
+        """Update dissimilarity map after finding a cherry.
+
+        Updates the dissimilarity map after joining together two nodes (m1, m2)
+        at a cherry m. For all other nodes v, the new dissimilarity map d' is:
+
+        d'(m, v) = (|m1| * d(m1, v) + |m2| * d(m2, v))/(|m1| + |m2|)
+
+        where |m1| is the size of cluster m1, i.e. the number of sample leaves
+        under node m1.
+
+        Args:
+            dissimilarity_map: A dissimilarity map to update
+            cherry: A tuple of indices in the dissimilarity map that are joining
+            new_node: New node name, to be added to the new dissimilarity map
+
+        Returns:
+            A new dissimilarity map, updated with the new node
+        """
+        i_size, j_size = cluster_to_cluster_size[cherry[0]], cluster_to_cluster_size[cherry[1]]
+        cluster_to_cluster_size[new_node] = i_size + j_size
```

Comment:
Yes, here I think you can use a private variable stored in the `UPGMASolver`. You can instantiate the mapping in the `__init__` function, I think.